### ISSUE TITLE
Java DSL

### ISSUE THREAD

**Issue by smartoma** [takari/polyglot-maven#28]
Do you have any plans to add Java DSL as an alternative language for pom?

**Comment by huntc** [comment]
PRs are welcome. :-)

I'd recommend Java 8. The class path for polyglot is different to your project so Java 8 would be fine.

**Comment by iirekm** [comment]
Java is quite clumsy as for DSLs, why not use Groovy DSL? Syntax of Groovy is quite similar to Java's

**Comment by paulvi** [comment]
There is already Groovy module https://github.com/takari/maven-polyglot/tree/master/polyglot-groovy
and no one for Java.

Personally I think lacks JSON-like object definitions, that are in JavaScript #16

**Comment by iirekm** [comment]
By 'Groovy DSL' I just meant polyglot-groovy :-) writing a DSL in Java makes no big sense

**Comment by sveryovka** [comment]
Can someone propose good syntax for Java DSL? I mean not just how to add dependencies but to cover at least all of the major functionality. If it will look good I can implement it.

**Comment by paulvi** [comment]
0. Shouldn't Java flavour just import directly Mode, Dependency, Plugin etc classes.
That may be to verbose, but the flavor would be great tool for those creating maven-plugins, IMHO

1. Isn't json to become part of Java ?

As of now, I think Groovy is OK, see Java/Groovy example #95

What I would like to improve though is have plugin configuration not so deep inside.
@sveryovka Would variable be supported?

2. How about some method invocation like 

```java
addPlugin("maven-release-plugin", new ImmutableMap.Builder<String, Integer>()
           .put("preparationGoals", "clean install")
           .put("mavenExecutorId", "forked-path")
           .build()
)
```

passing configuration is still awkward.

**Comment by obfischer** [comment]
Hi, if you consider a DSL for Java, you should have a look at this blog post. The author uses some features of JDK 8 to develop a quite clean DSL to generate HTML and Puppet files.

http://benjiweber.co.uk/blog/2015/08/21/html-in-java/

**Comment by paulvi** [comment]
@obfischer Oliver, that is great example, so everything is either method or lamda

```java
public interface JavaMavenDsl extends
  Project,
  GAV,
  Licenses,
  Modules,
  CSM,
  Properties,
  DependencyManagement
```

```java
project(
  modelVersion -> "4.0.0",
  parent(
    groupId -> "io.takari",
    artifactId -> "takari",
    version -> "14"
  ),
  groupId -> "io.takari.polyglot",
  artifactId -> "polyglot",
  version -> "0.1.16-SNAPSHOT",
  packaging -> "pom",
  name -> "Polyglot :: Aggregator",
  licenses(
    license(
      name -> "The Eclipse Public License, Version 1.0",
      url -> "http://www.eclipse.org/legal/epl-v10.html",
      distribution -> "repo"
    )    
  )
  ...
  plugin("maven-release-plugin", new ImmutableMap.Builder<String, String>()
           .put("preparationGoals", "clean install")
           .put("mavenExecutorId", "forked-path")
           .build()
)
);

```

**Comment by paulvi** [comment]
Example above went to blog https://dzone.com/articles/maven-polyglot-replacing-a-pom-with-a-scala-or-groovy-build-script

**Comment by sveryovka** [comment]
Hi,

I have experimented with what is possible to achive with Java DSL. You can find my edperiment at https://github.com/sveryovka/polyglot-maven/tree/java-dsl 
It is far from complete. Even dynamic compilation when running from command line is failing so you need to run all samples from IDE. But it lets you to see of what what are the different possible syntax options. 

For example 
https://github.com/sveryovka/polyglot-maven/blob/java-dsl/polyglot-java/src/main/java/org/sonatype/maven/polyglot/java/test/ModelTest.java
And to be honest at this point I am not sure I like any of the options. For example, let's look at potential syntax to define resource:
A:
`resource(			
    directory -> "c://foodir",
    filtering -> "true",
    targetPath -> "c://bardir",
    includes("*.a"),
    excludes("*.b")
)` looks ok but IDE autocompietion will not help you to write this code which is bad.
B:
`resource()
.directory("c://foodir").filtering(true).targetPath("c://bardir").includes("*.a").excludes("*.b")
.buildResource(),`
IDE auto completion works but it does not look as good.
C:
`resource(r -> {r.directory="c://foodir"; r.filtering=true; r.targetPath="c://bardir"; r.includes="*.a"; r.excludes="*.b";}) 
` I am not sure if I like the look of it and IDE auto completion works only when you inside of lambda body {}

**Comment by sveryovka** [comment]
I have updated my experiment branch with code to dynamically compile java.pom file. It makes it easier to play with it.

**Comment by mosabua** [comment]
Cool ... should we merge this into the main codebase .. even if its experimental at the moment?

**Comment by sveryovka** [comment]
I have published PR. There are many limitations and I expect bugs. But if we can get more feedback on syntax by merging it - it may be a good thing to do.

**Comment by paulvi** [comment]
@sveryovka wrote 

> but IDE autocompietion will not help you to write this code which is bad.

maven model should not be on project classpath, so default Java Editor can't and should not do auto compietion 


Now PR #149 is merged and closed, and it is unlikely that ultimate implementation will follow soon,  
this issue can be continued as proposed experimental code already within this codebase.

**Comment by sveryovka** [comment]
As for me ease of use and convenience of use are most important things that may drive creation of any new DSL. And lack of auto completion and any other assistance from IDE is a big problem from this point of view. 

I suppose you mean "production ready" implementation when you say "ultimate implementation". Code is new and definitely requires significant testing. This requires people who use it and test it. At this point to use the code and even to improve it you don't need to know maven internals. So anyone can help. Up until now my motivation was to see what syntax is possible and how it will look for a complex objects like "build" and "plugin". Now it is not that hard to contribute by testing it and even publishing PRs.

**Comment by mosabua** [comment]
The DSL is now in master and will be released with the upcoming 0.3.0 release. As such this issue is fixed. Of course further work can be done via PRs...
